src/metabase/data_apps/models.clj

```diff
@@ -0,0 +1,128 @@
+(ns metabase.data-apps.models
+  (:require
+   [metabase.models.interface :as mi]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(methodical/defmethod t2/table-name :model/App           [_model] :app)
+(methodical/defmethod t2/table-name :model/AppDefinition [_model] :app_definition)
+(methodical/defmethod t2/table-name :model/AppRelease    [_model] :app_release)
+
+(doseq [model [:model/App
+               :model/AppDefinition
+               :model/AppRelease]]
+  (derive model :metabase/model))
+
+(derive :model/App :hook/timestamped?)
+(derive :model/AppDefinition :hook/created-at-timestamped?)
+(derive :model/AppDefinition :hook/entity-id)
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                         :model/App                                             ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(t2/deftransforms :model/App
+  {:config mi/transform-json})
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                     :model/AppDefinition                                       ;;
+;;-----------------------------------------------------------------------------------------------;;
+
+(mr/def ::AppDefinitionConfig
+  [:maybe [:map {:closed true}
+           [:version pos-int?]]])
+
+(defn- validate-app-definition
+  "Validate an AppDefinition."
+  [app-definition]
+  (mu/validate-throw ::AppDefinitionConfig (:config app-definition)))
+
+(defn- next-version
+  [app-id]
+  (or (t2/select-one-fn :version :model/AppDefinition :app_id app-id {:order-by [[:version :desc]]
+                                                                      :limit 1})
+      0))
```

Comment:
We'll want to calculate the new version within the SQL statement that inserts new version, for atomicity.